index.d.ts

```diff
@@ -0,0 +1,456 @@
+declare module 'oidc-provider' {
+    import { Request, Response, Handler } from 'express';
+    import Koa = require('koa');
+    import Router = require('koa-router');
+    import { IncomingMessage, ServerResponse } from 'http';
+    import EventEmitter = require('events');
+    import { JWK } from 'node-jose';
+
+    type SubjectType = 'public' | 'parwise';
+
+    interface IClient {
+        applicationType?: 'web' | 'native';
+        grant_types?: string[];
+        id_token_signed_response_alg?: string;
+        require_auth_time?: boolean;
+        response_types?: string[];
+        subject_type?: SubjectType;
+        token_endpoint_auth_method?: string;
+        request_uris?: string[];
+        client_id?: string;
+        client_secret?: string;
+        redirect_uris: string[];
+        introspection_endpoint_auth_method?: string;
+    }
+
+    interface IAdapter {
+        upsert: (id: string, payload: string, expiresIn: number) => Promise<void>;
+        find: (id: string) => Promise<void>;
+        findByUserCode: (userCode: string) => Promise<void>;
+        consume: (id: string) => Promise<void>;
+        destroy: (id: string) => Promise<void>;
+        connect?: (provider: Provider) => Promise<void>;
+        key: (id: string) => string;
```

Comment:
remove this, that's not part of the expected API, some example adapters just use this internally